rinja/src/ascii_str.rs

```diff
@@ -0,0 +1,297 @@
+        }
+        result
+    }
+
+    #[inline]
+    pub const fn from_slice(src: &[AsciiChar]) -> &Self {
+        // SAFETY: `Self` is transparent over `[AsciiChar]`.
+        unsafe { core::mem::transmute::<&[AsciiChar], &AsciiStr>(src) }
+    }
+
+    #[inline]
+    pub const fn as_str(&self) -> &str {
+        // SAFETY: `Self` has the same layout as `str`,
+        // and all ASCII characters are valid UTF-8 characters.
+        unsafe { core::mem::transmute::<&AsciiStr, &str>(self) }
+    }
+
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+// Must not implement `DerefMut`. Not every `char` is an ASCII character.
+impl Deref for AsciiStr {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_str()
+    }
+}
+
+impl<Idx> Index<Idx> for AsciiStr
+where
+    [AsciiChar]: Index<Idx, Output = [AsciiChar]>,
+{
+    type Output = [AsciiChar];
+
+    #[inline]
+    fn index(&self, index: Idx) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+impl<Idx> IndexMut<Idx> for AsciiStr
+where
+    [AsciiChar]: IndexMut<Idx, Output = [AsciiChar]>,
+{
+    #[inline]
+    fn index_mut(&mut self, index: Idx) -> &mut Self::Output {
+        &mut self.0[index]
+    }
+}
+
+impl Default for &'static AsciiStr {
+    #[inline]
+    fn default() -> Self {
+        // SAFETY: `Self` has the same layout as `str`.
+        unsafe { core::mem::transmute::<&str, &AsciiStr>("") }
+    }
+}
+
+impl AsciiChar {
+    #[inline]
+    pub const fn new(c: u8) -> Self {
+        if c.is_ascii() {
+            // SAFETY: We just checked that the input are `Ascii`,
+            // and `Self` has the same representation as `u8`.
+            unsafe { core::mem::transmute::<u8, AsciiChar>(c) }
+        } else {
+            panic!()
+        }
+    }
+
+    #[inline]
+    pub const fn slice_as_bytes<const N: usize>(src: &[AsciiChar; N]) -> &[u8; N] {
+        // SAFETY: `[AsciiChar]` has the same layout as `[u8]`.
+        unsafe { core::mem::transmute::<&[AsciiChar; N], &[u8; N]>(src) }
+    }
+
+    #[inline]
+    pub const fn two_digits(d: u32) -> [Self; 2] {
+        if d >= 100 {
+            panic!();
+        }
+
+        match core::str::from_utf8(&[b'0' + (d / 10) as u8, b'0' + (d % 10) as u8]) {
+            Ok(s) => AsciiStr::new_sized(s),
+            Err(_) => panic!(),
+        }
+    }
+
+    #[inline]
+    pub const fn two_hex_digits(d: u32) -> [Self; 2] {
+        const HEX: &[u8; 16] = b"0123456789abcdef";
+
+        if d >= 256 {
+            panic!();
+        }
+        match core::str::from_utf8(&[HEX[(d / 16) as usize], HEX[(d % 16) as usize]]) {
+            Ok(s) => AsciiStr::new_sized(s),
+            Err(_) => panic!(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(u8)]
+pub enum AsciiChar {
```

Comment:
thought: I think `AsciiChar` could just be `AsciChar(u8)` and it would not be less safe. But this enum does appear to be tested well.

Reply:
You are right. This also allowed me to encapsulated the type further and replace two `transmute`s.